src/java/org/apache/cassandra/cql3/statements/TransactionStatement.java

```diff
@@ -177,9 +178,16 @@ TxnNamedRead createNamedRead(NamedSelect namedSelect, QueryOptions options)
         SinglePartitionReadQuery.Group<SinglePartitionReadCommand> selectQuery = (SinglePartitionReadQuery.Group<SinglePartitionReadCommand>) readQuery;
 
         if (selectQuery.queries.size() != 1)
-            throw new IllegalArgumentException("Within a transaction, SELECT statements must select a single partition; found " + selectQuery.queries.size() + " partitions");
+        {
+            if (!TxnDataName.returning().equals(namedSelect.name))
+                throw new IllegalArgumentException("Within a transaction, SELECT statements must select a single partition; found " + selectQuery.queries.size() + " partitions");
+            // multi partitions on the same table are only allowed for the returning clause
+            return IntStream.range(0, selectQuery.queries.size())
+                            .mapToObj(i -> new TxnNamedRead(TxnDataName.returning(i), selectQuery.queries.get(i)))
+                            .collect(Collectors.toList());
```

Comment:
This should be a pretty hot path, so I'd have two potential concerns.

1.) I think we'll create less overall garbage w/ something like...

```
List<TxnNamedRead> list = new ArrayList<>(selectQuery.queries.size());
for (int i = 0; i < selectQuery.queries.size(); i++)
    list.add(new TxnNamedRead(TxnDataName.returning(i), selectQuery.queries.get(i)));
```

2.) Given how many times we'll only have one query, I'd leave `createNamedRead()` alone and create a new `createNamedReads()` to house the new logic here. No point in creating even a singleton list if we don't have to.